transport/src/main/java/io/netty/channel/local/LocalChannel.java

```diff
@@ -246,34 +246,42 @@ protected void doClose() throws Exception {
-                            peer.tryClose(peerIsActive);
+            try {
+                // Need to execute the close in the correct EventLoop (see https://github.com/netty/netty/issues/1777).
+                // Also check if the registration was not done yet. In this case we submit the close to the EventLoop
+                // to make sure its run after the registration completes
+                // (see https://github.com/netty/netty/issues/2144).
+                EventLoop peerEventLoop = peer.eventLoop();
+                final boolean peerIsActive = peer.isActive();
+                if (peerEventLoop.inEventLoop() && !registerInProgress) {
+                    peer.tryClose(peerIsActive);
+                } else {
+                    try {
+                        peerEventLoop.execute(new Runnable() {
+                            @Override
+                            public void run() {
+                                peer.tryClose(peerIsActive);
+                            }
+                        });
+                    } catch (Throwable cause) {
+                        logger.warn("Releasing Inbound Queues for channels {}-{} because exception occurred!",
+                                this, peer, cause);
+                        if (peerEventLoop.inEventLoop()) {
+                            peer.releaseInboundBuffers();
+                        } else {
+                            // inboundBuffers is a SPSC so we may leak if the event loop is shutdown prematurely or
+                            // rejects the close Runnable but give a best effort.
+                            peer.close();
                         }
-                    });
-                } catch (Throwable cause) {
-                    logger.warn("Releasing Inbound Queues for channels {}-{} because exception occurred!",
-                            this, peer, cause);
-                    releaseInboundBuffers();
-                    if (peerEventLoop.inEventLoop()) {
-                        peer.releaseInboundBuffers();
-                    } else {
-                        // inboundBuffers is a SPSC so we may leak if the event loop is shutdown prematurely or rejects
-                        // the close Runnable but give a best effort.
-                        peer.close();
+                        PlatformDependent.throwException(cause);
                     }
-                    PlatformDependent.throwException(cause);
                 }
+            } finally {
+                // We need to release all the buffers that may be put into our inbound queue since we closed the Channel
+                // to ensure we not leak any memory. This is fine as it basically gives the same guarantees as TCP which
+                // means even if the promise was notified before its not really guaranteed that the "remote peer" will
+                // see the buffer at all.
+                releaseInboundBuffers();
```

Comment:
Feel free to refactor but this is what I was thinking of:

```java
    @Override
    protected void doClose() throws Exception {
        final LocalChannel peer = this.peer;
        boolean stateChangedToClosed = state != State.CLOSED;
        try {
            if (stateChangedToClosed) {
                // Update all internal state before the closeFuture is notified.
                if (localAddress != null) {
                    if (parent() == null) {
                        LocalChannelRegistry.unregister(localAddress);
                    }
                    localAddress = null;
                }

                // State change must happen before finishPeerRead to ensure writes are released either in doWrite or
                // channelRead.
                state = State.CLOSED;

                // Preserve order of event and force a read operation now before the close operation is processed.
                finishPeerRead(this);

                ChannelPromise promise = connectPromise;
                if (promise != null) {
                    // Use tryFailure() instead of setFailure() to avoid the race against cancel().
                    promise.tryFailure(DO_CLOSE_CLOSED_CHANNEL_EXCEPTION);
                    connectPromise = null;
                }
            }

            if (peer != null) {
                this.peer = null;
                // Need to execute the close in the correct EventLoop (see https://github.com/netty/netty/issues/1777).
                // Also check if the registration was not done yet. In this case we submit the close to the EventLoop
                // to make sure its run after the registration completes
                // (see https://github.com/netty/netty/issues/2144).
                EventLoop peerEventLoop = peer.eventLoop();
                final boolean peerIsActive = peer.isActive();
                if (peerEventLoop.inEventLoop() && !registerInProgress) {
                    peer.tryClose(peerIsActive);
                } else {
                    try {
                        peerEventLoop.execute(new Runnable() {
                            @Override
                            public void run() {
                                peer.tryClose(peerIsActive);
                            }
                        });
                    } catch (Throwable cause) {
                        logger.warn("Releasing Inbound Queues for channels {}-{} because exception occurred!",
                                this, peer, cause);
                        if (peerEventLoop.inEventLoop()) {
                            peer.releaseInboundBuffers();
                        } else {
                            // inboundBuffers is a SPSC so we may leak if the event loop is shutdown prematurely or
                            // rejects the close Runnable but give a best effort.
                            peer.close();
                        }
                        PlatformDependent.throwException(cause);
                    }
                }
            }
        } finally {
            if (stateChangedToClosed) {
                // We need to release all the buffers that may be put into our inbound queue since we closed the Channel
                // to ensure we not leak any memory. This is fine as it basically gives the same guarantees as TCP which
                // means even if the promise was notified before its not really guaranteed that the "remote peer" will
                // see the buffer at all.
                releaseInboundBuffers();
            }
        }
    }
```